lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,604 @@
+    throw new ERR_MODULE_NOT_FOUND(
+      resolved.pathname, fileURLToPath(base), 'module');
+  }
+
+  if (resolved.pathname.endsWith('/')) return resolved;
+  const path = fileURLToPath(resolved);
+
+  if (!tryStatSync(path).isFile()) {
+    throw new ERR_MODULE_NOT_FOUND(
+      path || resolved.pathname, fileURLToPath(base), 'module');
+  }
+
+  return resolved;
+}
+
+function throwExportsNotFound(subpath, packageJSONUrl, base) {
+  throw new ERR_PACKAGE_PATH_NOT_EXPORTED(
+    fileURLToPath(packageJSONUrl), subpath, fileURLToPath(base));
+}
+
+function throwSubpathInvalid(subpath, packageJSONUrl, base) {
+  throw new ERR_INVALID_MODULE_SPECIFIER(
+    fileURLToPath(packageJSONUrl), subpath, fileURLToPath(base));
+}
+
+function throwExportsInvalid(
+  subpath, target, packageJSONUrl, base) {
+  if (typeof target === 'object' && target !== null) {
+    target = JSONStringify(target, null, '');
+  } else if (ArrayIsArray(target)) {
+    target = `[${target}]`;
+  } else {
+    target = `${target}`;
+  }
+  throw new ERR_INVALID_PACKAGE_TARGET(
+    fileURLToPath(packageJSONUrl), null, subpath, target, fileURLToPath(base));
+}
+
+function resolveExportsTargetString(
+  target, subpath, match, packageJSONUrl, base) {
+  if (target[0] !== '.' || target[1] !== '/' ||
+      (subpath !== '' && target[target.length - 1] !== '/')) {
+    throwExportsInvalid(match, target, packageJSONUrl, base);
+  }
+
+  const resolved = new URL(target, packageJSONUrl);
+  const resolvedPath = resolved.pathname;
+  const packagePath = new URL('.', packageJSONUrl).pathname;
+
+  if (!resolvedPath.startsWith(packagePath) ||
+      resolvedPath.includes('/node_modules/', packagePath.length - 1)) {
+    throwExportsInvalid(match, target, packageJSONUrl, base);
+  }
 
-function defaultResolve(specifier, { parentURL } = {}, defaultResolve) {
+  if (subpath === '') return resolved;
+  const subpathResolved = new URL(subpath, resolved);
+  const subpathResolvedPath = subpathResolved.pathname;
+  if (!StringPrototypeStartsWith(subpathResolvedPath, resolvedPath) ||
+      StringPrototypeIncludes(subpathResolvedPath,
+                              '/node_modules/', packagePath.length - 1)) {
+    throwSubpathInvalid(match + subpath, packageJSONUrl, base);
+  }
+  return subpathResolved;
+}
+
+function isArrayIndex(key) {
+  const keyStr = `${key}`;
+  if (`${+keyStr}` !== keyStr) return false;
+  key >>>= 0;
+  if (ObjectIs(key, 0)) return true;
+  return key > 0 && key < 0xFFFF_FFFF;
+}
+
+function resolveExportsTarget(
+  packageJSONUrl, target, subpath, packageSubpath, base) {
+  if (typeof target === 'string') {
+    const resolved = resolveExportsTargetString(
+      target, subpath, packageSubpath, packageJSONUrl, base);
+    return finalizeResolution(resolved, base);
+  } else if (ArrayIsArray(target)) {
+    if (target.length === 0)
+      throwExportsInvalid(packageSubpath, target, packageJSONUrl, base);
+
+    let lastException;
+    for (const targetItem of target) {
```

Comment:
is this to avoid the iterator prototype?

Reply:
yes, since it is mutable/deletable

Reply:
Is there a fundamental difference between mutating the iterator prototype and defining numeric properties on `Array.prototype`? In both cases it would be possible to affect what is returned here afaict.

Reply:
Yes, the fundamental difference is that the numeric properties on `Array.prototype` represent the contents of the array (ie, if `target` is a sparse array, they'd show up, but only then) but `Symbol.iterator` is an implementation detail of "how node iterates the array", one that shouldn't be observable.

Reply:
@jkrems yes, numeric properties on arrays *and* .length are own properties. Holey arrays have their own hell per https://docs.google.com/presentation/d/11fkQeEisoszNGF8SrautVT1ltSnsQBWRxJ4usoc-g_o/edit?usp=sharing , but here since we know the arrays are not holey, it isn't a problem